### PR TITLE
(PUP-4561) Extract platform-specific code

### PR DESCRIPTION
**Pull request by MikaelSmith**
Before this commit, the `prepare` class was full of platform-specific
code that needed to be refactored before non-RedHat platforms could be
supported.

Extract the platform-specific code to `params` and a new class that is
selected based on the `osfamily` fact.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
This will definitely have a merge conflict with #11

File: manifests/osfamily/redhat.pp [hunk 1]
```diff
@@ -0,0 +1,45 @@
+class agent_upgrade::osfamily::redhat {
+  #  if $caller_module_name != $module_name {
+  #  fail("Use of private class ${name} by ${caller_module_name}")
+  #}
```
Comment (joshcooper) [inline manifests/osfamily/redhat.pp]: can be removed?

Comment (MikaelSmith) [inline manifests/osfamily/redhat.pp]: Oh right, I was going to try the guard I saw in another repo. I'll check whether I want to enable it or remove it completely.

File: manifests/prepare.pp [hunk 2]
```diff
@@ -53,15 +54,25 @@
...
 
-  # Install PC1 yum repo; based off puppetlabs_yum
-  contain '::agent_upgrade::puppetlabs_yum'
+  # PLATFORM SPECIFIC CONFIGURATION
+  # Break out the platform-specific configuration into subclasses, dependent on
+  # the osfamily of the client being configured.
+
+  case $::osfamily {
+    'redhat', 'debian', 'windows', 'solaris', 'aix', 'suse': {
+      contain downcase("::agent_upgrade::osfamily::${::osfamily}")
```
Comment (joshcooper) [inline manifests/prepare.pp]: Is this a common module paradigm for "contain"ing platform specific functionality?

Comment (MikaelSmith) [inline manifests/prepare.pp]: It's one I borrowed from the pe-agent module.

**Pull request merged**